File: app/src/main/java/com/qurantv/app/data/cache/JsonDiskCache.kt
package com.qurantv.app.data.cache

import android.content.Context
import java.io.File
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

/**
 * Simple on-disk JSON cache (PROMPT.md Part 8):
 *  - catalog responses: TTL 24h
 *  - timing data: immutable → cached forever, keyed per (read, surah)
 *  - quran text (Quran.com): cached per surah forever
 *
 * Writes are atomic (tmp file + rename). [singleFlight] prevents duplicate
 * network calls for the same key.
 */
class JsonDiskCache(private val context: Context) {

    private val root = File(context.filesDir, "json_cache")
    private val locks = ConcurrentHashMap<String, Mutex>()

    suspend fun read(category: String, key: String, maxAgeMs: Long? = null): String? =
        withContext(Dispatchers.IO) {
            val file = file(category, key)
            if (!file.exists()) return@withContext null
            if (maxAgeMs != null && System.currentTimeMillis() - file.lastModified() > maxAgeMs) {
                return@withContext null
            }
            runCatching { file.readText() }.getOrNull()
        }

    suspend fun write(category: String, key: String, content: String): Unit =
        withContext(Dispatchers.IO) {
            val file = file(category, key)
            file.parentFile?.mkdirs()
            val tmp = File(file.parentFile, file.name + ".tmp")
            tmp.writeText(content)
            if (!tmp.renameTo(file)) {
                // rename can fail on some devices; fall back to direct write
                file.writeText(content)
                tmp.delete()
            }
        }

    fun file(category: String, key: String): File =
        File(File(root, category), safeKey(key))

    private fun safeKey(key: String): String {
        val sanitized = key.replace(Regex("[^A-Za-z0-9._-]"), "_")
        return sanitized.takeLast(120)
    }

    /** Serializes concurrent loads of the same key (single-flight per key). */
    suspend fun <T> singleFlight(flightKey: String, block: suspend () -> T): T {
        val mutex = locks.computeIfAbsent(flightKey) { Mutex() }
        return mutex.withLock { block() }
    }

    companion object {
        const val TTL_24H: Long = 24 * 60 * 60 * 1000L
        const val CATALOG = "catalog"
        const val TIMING = "timing"
        const val QURAN_TEXT = "quran_text"
    }
}

File: web/src/data/cache/JsonDiskCache.ts
// Ported from app/src/main/java/com/qurantv/app/data/cache/JsonDiskCache.kt
// localStorage-backed JSON cache:
//  - catalog responses: TTL 24 h
//  - timing data: immutable → cached forever, keyed per (read, surah)
//  - quran text (Quran.com): cached per surah forever
// singleFlight prevents duplicate network calls for the same key.
//
// localStorage quota is ~5-10 MB on TV webviews; timing + text data is small
// (per-surah timings are a few KB; 114 surahs ≈ ~1 MB), so this is fine.

const TTL_24H = 24 * 60 * 60 * 1000;

export class JsonDiskCache {
  private readonly root = "qurantv_json_cache";
  private readonly flights = new Map<string, Promise<unknown>>();

  /** Raw stored value, or null when missing/expired/unparseable. */
  async read(category: string, key: string, maxAgeMs: number | null = null): Promise<string | null> {
    try {
      const raw = localStorage.getItem(this.storageKey(category, key));
      if (raw === null) return null;
      const parsed = JSON.parse(raw) as { v: string; t: number };
      if (maxAgeMs !== null && Date.now() - parsed.t > maxAgeMs) return null;
      return parsed.v;
    } catch {
      return null;
    }
  }

  async write(category: string, key: string, content: string): Promise<void> {
    try {
      localStorage.setItem(
        this.storageKey(category, key),
        JSON.stringify({ v: content, t: Date.now() }),
      );
    } catch {
      // quota exceeded — drop silently (cache is best-effort)
    }
  }

  /** Serializes concurrent loads of the same key (single-flight per key). */
  async singleFlight<T>(flightKey: string, block: () => Promise<T>): Promise<T> {
    const existing = this.flights.get(flightKey) as Promise<T> | undefined;
    if (existing) return existing;
    const p = block().finally(() => {
      this.flights.delete(flightKey);
    });
    this.flights.set(flightKey, p);
    return p;
  }

  private storageKey(category: string, key: string): string {
    return `${this.root}/${category}/${this.safeKey(key)}`;
  }

  private safeKey(key: string): string {
    const sanitized = key.replace(/[^A-Za-z0-9._-]/g, "_");
    return sanitized.slice(-120);
  }
}

export const CACHE_TTL_24H = TTL_24H;

export const CACHE = {
  CATALOG: "catalog",
  TIMING: "timing",
  QURAN_TEXT: "quran_text",
  KSU_HILITES: "ksu_hilites",
} as const;
